[pull] primary from flybywiresim:primary (#8)
Valastiri/fbw-userguide: docs/a32nx-dev/overview.md

# FlyByWire Development Guide Overview

## Introduction

> This guide has the objective of teaching you how to contribute efficiently to the FlyByWire A32MX project.

Reviewer: ```suggestion
This guide has the objective of teaching you how to contribute efficiently to the FlyByWire A32NX project.
```

You will learn how to:

> - Setup a functional development environment;
> - Get familiar with the technologies involved;
> - Understand our build process;
> - Integrate into our development cycle and best practices.

Reviewer: No need for a ; at the end of the line
```suggestion
- Setup a functional development environment
- Get familiar with the technologies involved
- Understand our build process
- Integrate into our development cycle and best practices.
```

## Setting up a development environment

> Working on the addon will require, at a minimum, the following software:

Reviewer: ```suggestion
Working on the addon will require at a minimum the following software:
```

- A code editor (with at least syntax highlighting, search and replace, and support for workspaces)
- A `git` client (along with any editor integrations of your choice)

> - Docker desktop (either WSL2 or HyperV backend work, but the latter is faster)

Reviewer: ```suggestion
- Docker Desktop (either WSL2 or HyperV backend work, but the latter is faster)
```
the official name is Docker Desktop, so should be capitalized

> - Docker desktop (either WSL2 or HyperV backend work, but the latter is faster)

Reviewer: Link to some of these

Most of our team works with either Visual Studio Code or IntelliJ IDEA-based IDEs for development. Obviously, your choice is yours as long as the resulting code conforms to our standards.

## Used technologies

Our tech stack includes the following:

> - Systems development for aircraft uses Rust and the `msfs-rs` library

Reviewer: ```suggestion
- Systems development for aircraft using Rust and the `msfs-rs` library
```

- Avionics programming is done using JavaScript or TypeScript (depending on the project), with the `React.js` rendering library.
- Front-end web or desktop app development uses the same technologies outlined above

> - Server-side development uses `nest.js` for the API and `MySQL` for the database backend

Reviewer: Again, should link to some things

Knowledge of all items on this list is obviously not necessary, but this can hopefully give you a better insight into what your skills can fit into.

## Build process

> Due to how Microsoft Flight Simulator package building works, it is necessary for us to maintain our own, bespoke build tools.

Reviewer: ```suggestion
Due to how Microsoft Flight Simulator package building works, it is necessary for us to maintain our own bespoke build tools.
```

The high-level objective involving the build process is to eventually have all code in the `src/` directory.

> Commonly, the following items will be found in the source:

Reviewer: Maybe rephrase this?

```
- src/
    - instruments/                  Avionics code built using the rollup config
        - rollup.config.js
    - behavior/                     Compiled and verified ModelBehavior code
    - model/                        Model source files to be merged at build time
    - <directories for systems>     Rust/C++ source code for systems
    ... more to come
```

> Each directory will usually contain a build.js/build.sh file that is called from a central build script located in `srcipts/build.sh`.

Reviewer: ```suggestion
Each directory will usually contain a build.js/build.sh file that is called from a central build script located in `scripts/build.sh`.
```

A Docker container is used to ship pre-configured build environments to developers. It contains necessary library headers and compilers, and requires no user intervention at all.

> This build process is optimized to be run in a cloud CI pipelines on GitHub Actions. This allows us to automatically distribute development master builds to thousands of users without any manual intervention. It lets the team deploy critical fixes to production in a matter of minutes.

Reviewer: ```suggestion
This build process is optimized to be run in cloud CI pipelines on GitHub Actions. This allows us to automatically distribute development master builds to thousands of users without any manual intervention. It lets the team deploy critical fixes to production in a matter of minutes.
```

The automatic CI pipeline is sometimes manually added to other branches in order to provide the community with experimental versions.

An automatic system that allows us to enable CI for specific branches might be considered in the future.

## Development practices

FlyByWire Simulations aims to maintain professional standards as well as industry-wide best practices in order to maximize developer productivity and deliver high-quality, tested products.

This is why we adopt a set of rules and practices that create a strong framework, not only around the usage of `git` but also surrounding code quality.

### Version control

Our version control approach maintains the following guidelines:

> - Commit messages should be short and to the point. The Angular commit message convention is often used and recommended but not enforced at the moment

Reviewer: Should probably link to the Angular commit message convention

- Merge commits are discouraged in commit histories and disallowed when merging pull requests
- Squash commits are exclusively used for pull request merges  
- One approving review from a maintainer is required to merge a pull request
- Positive reports from a number of QA testers (varies depending on the PR scope and size) are required before merging, unless a maintainer judges appropriate to bypass this rule
- Creation of draft pull requests is required when a developer wishes to get working on a project, to avoid duplicate PR situations

### Code quality

Code quality is of an utmost importance as it allows other developers to more easily find bugs and understand logic.

An ESLint configuration is provided and should cover most code style rules.

A few guidelines apply:

- Prefer self-documenting code over having to write tons of comments
- However, do not hesitate to write comments when complex logic is used

> - Use language-specific docstring tools (JSDoc, rustdoc)

Reviewer: maybe link to these?

- Maintain separation of concerns
- Do not be afraid to separate code into logical blocks, even if resulting sub-functions are only used locally (while maintaining a reasonable limit)

> - Avoid polling - prefer reactive APIs (but remain sure of not introducing state that cannot be externally read)

Reviewer: ```suggestion
- Avoid polling - prefer reactive APIs (but make sure you're not introducing states that cannot be externally read)
```

Reviewer 2: state should stay singular as we're talking about state in general

- Avoid introducing new technologies/languages without talking with the development team

Additionally, a few practices are straight up forbidden:

- Do not use `async/await` without a good reason (it does not make code run faster unless you are calling threaded APIs like `fetch`)
- Do not use hungarian notation (b, m, g identifier prefixes)

> - Do not perform business logic calculations in presentation code

Reviewer: I don't quite understand this line

### Reference material

> Creating a high-fidelity aircraft is a serious task that requires focus and dedication. But more importantly, it's a strive for accuracy and realism. Therefore, any change that affects the appearance or behavior of a plane needs to reference IRL documentation.

Reviewer: ```suggestion
Creating a high-fidelity aircraft is a serious task that requires focus and dedication. But more importantly, it's a strive for accuracy and realism. Therefore, any change that affects the appearance or behavior of a plane needs to reference real documentation.
```
IRL doesn't sound professional
